### PR TITLE
planner: fix index join on unmatched collation suffix columns paniced

### LINKED ISSUE #24706: open new collation + index join + not null column，occur error: index out of  range
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
set  new_collations_enabled_on_first_bootstrap: true

```
CREATE TABLE `t1` (
  `c_int` int(11) NOT NULL,
  `c_str` varbinary(40) NOT NULL,
  `c_datetime` datetime DEFAULT NULL,
  `c_timestamp` timestamp NULL DEFAULT NULL,
  `c_double` double DEFAULT NULL,
  `c_decimal` decimal(12,6) DEFAULT NULL,
  PRIMARY KEY (`c_int`,`c_str`) /*T![clustered_index] CLUSTERED */,
  KEY `c_int_2` (`c_int`),
  KEY `c_decimal` (`c_decimal`),
  KEY `c_datetime` (`c_datetime`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH( `c_int` )
PARTITIONS 4;

CREATE TABLE `t2` (
  `c_int` int(11) NOT NULL,
  `c_str` varchar(40) NOT NULL,
  `c_datetime` datetime DEFAULT NULL,
  `c_timestamp` timestamp NULL DEFAULT NULL,
  `c_double` double DEFAULT NULL,
  `c_decimal` decimal(12,6) DEFAULT NULL,
  PRIMARY KEY (`c_int`,`c_str`) /*T![clustered_index] CLUSTERED */,
  KEY `c_str` (`c_str`(2)),
  KEY `c_datetime` (`c_datetime`),
  KEY `c_timestamp` (`c_timestamp`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE ( `c_int` ) (
  PARTITION `p0` VALUES LESS THAN (10),
  PARTITION `p1` VALUES LESS THAN (20),
  PARTITION `p2` VALUES LESS THAN (30),
  PARTITION `p3` VALUES LESS THAN (MAXVALUE)
);
insert into t1 values(3,'nice mccarthy','2020-02-11 14:30:12','2020-04-09 13:26:39','93.496212','2.848000');
select * from t1 partition (p3), t2 partition (p3) where t1.c_int = t2.c_int and t1.c_str = t2.c_str;
```
### 2. What did you expect to see? (Required)

select return successful

### 3. What did you see instead (Required)

select return fail

mysql> select * from t1 partition (p3), t2 partition (p3) where t1.c_int = t2.c_int and t1.c_str = t2.c_str;
ERROR 1105 (HY000): runtime error: index out of range [1] with length 1

### 4. What is your TiDB version? (Required)

Release Version: v5.0.1
Edition: Community
Git Commit Hash: 1145e347d3469d8e89f88dce86f6926ca44b3cd8
Git Branch: heads/refs/tags/v5.0.1
UTC Build Time: 2021-04-23 05:51:17
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: close #24706 

Problem Summary:

index join on a, b, but b's collation is not matched, b's curIdxOff2KeyOff not be set to -1?

### What is changed and how it works?

What's Changed, How it Works:

index join on a, b, but b's collation is not matched, we should only build range (a)

### Related changes

- Need to cherry-pick to the release branch 5.0

### Check List 

Tests 

- Unit test

Side effects

- n/a

### Release note 

- Fix panic caused by index join on unmatch collation columns

### CODE REVIEW

**Comment by lysu** [comment]
@wjhuang2016  PTAL, I'm not very sure about this change - -

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -1272,6 +1272,7 @@ func (ijHelper *indexJoinBuildHelper) removeUselessEqAndInFunc(idxCols []*expres
 				ijHelper.curPossibleUsedKeys = append(ijHelper.curPossibleUsedKeys, idxCols[idxColPos])
 				continue
 			}
+			ijHelper.curIdxOff2KeyOff[idxColPos] = -1
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go:1275]: Why we did this change? Or should we strengthen the case to `STRING` type only?

Comment (lysu) [reply planner/core/exhaust_physical_plans.go:1275]: if we didn't change this

for table `t1 (a int not null, b varcharbinary not null,index (a, b))` and `t2(a int not null, b varchar not null, index(a, b))`

the query: 't1 join t2 on t1. a = tb2.a and t2.a = t2.b'

will cause `ranges` only contain column `a`'s value(caused by https://github.com/pingcap/tidb/pull/21201), but `keyOff2IdxOff` include both `a` and `b`'s offset in

https://github.com/pingcap/tidb/blob/bfdad7c75210a30c796d82ccd882f381c2c20478/executor/builder.go#L3677

so the query will panic

@winoros is any other advise to fix this~? I'm not familiar with those part logic 😂

**Comment by lysu** [comment]
@winoros PTAL if free thx...

**Comment by winoros** [comment]
The logic around this function is:
- We want to build an index's range using the join conditions.
- There are two type expression: `col_1=col_2`, `col_1=const, col1 in (const_list)`.
- So for a join, the expression used to construct ranges may be like `idx_col1=col_1 and idx_col2 in (const_list) and idx_col3 = col3`.
- The `curIdxOff2KeyOff` records the map relation between `idx_col1` with `col1` and `idx_col3` with `col3`, so it might be the case `[0, -1, 2]`.
- The `curNotUsedIndexCols` records the expression `idx_col2 in (const_list)`. So it would be [idx_col2] for this case.

The `removeUselessEqAndInFunc` is to maintain the `curIdxOff2KeyOff`. It will check the index column whose position is occupied by `-1` in `curIdxOff2KeyOff` can be found in `curNotUsedIndexCols`. Otherwise there's neither `idx_col=another_col` nor `idx_col=const or idx_col in (const_list)` expression so we cannot build range for index from its position. So we are going to reset the value to -1 after it.

For this the issue to be fixed in the pr. The collation check for the column is a little later. We'd better add the check earlier at https://github.com/pingcap/tidb/blob/master/planner/core/exhaust_physical_plans.go#L1219.

**Comment by lysu** [comment]
@winoros  @wjhuang2016  PTAL if free thx~

**Review by winoros — APPROVED** [review]

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -1212,14 +1212,21 @@ func (cwc *ColWithCmpFuncManager) String() string {
...
-func (ijHelper *indexJoinBuildHelper) resetContextForIndex(innerKeys []*expression.Column, idxCols []*expression.Column, colLens []int) {
+func (ijHelper *indexJoinBuildHelper) resetContextForIndex(innerKeys []*expression.Column, idxCols []*expression.Column, colLens []int, outerKeys []*expression.Column) {
 	tmpSchema := expression.NewSchema(innerKeys...)
 	ijHelper.curIdxOff2KeyOff = make([]int, len(idxCols))
 	ijHelper.curNotUsedIndexCols = make([]*expression.Column, 0, len(idxCols))
 	ijHelper.curNotUsedColLens = make([]int, 0, len(idxCols))
 	for i, idxCol := range idxCols {
 		ijHelper.curIdxOff2KeyOff[i] = tmpSchema.ColumnIndex(idxCol)
 		if ijHelper.curIdxOff2KeyOff[i] >= 0 {
+			// not use join column if collation is unmatched if new collation enabled.
```
Comment (wjhuang2016) [inline planner/core/exhaust_physical_plans.go:1223]:
```suggestion
			// Don't use the join columns if their collations are unmatched and the new collation is enabled.
```

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**